bundles/core/org.eclipse.smarthome.core.scheduler/src/main/java/org/eclipse/smarthome/core/scheduler/common/ThreadPoolManager.java

```diff
@@ -217,6 +260,108 @@ protected void afterExecute(Runnable runnable, Throwable throwable) {
         }
     }
 
+    public static class ExpressionThreadPoolExecutor extends ScheduledThreadPoolExecutor {
+
+        private List<Runnable> running = Collections.synchronizedList(new ArrayList<Runnable>());
+        private Map<Runnable, Runnable> scheduled = Collections.synchronizedMap(new HashMap<Runnable, Runnable>());
+        private Map<Runnable, Future<?>> futures = Collections.synchronizedMap(new HashMap<Runnable, Future<?>>());
+
+        public ExpressionThreadPoolExecutor(final String poolName, int corePoolSize) {
+            this(poolName, corePoolSize, new NamedThreadFactory(poolName), new ThreadPoolExecutor.DiscardPolicy() {
+                // The pool is bounded and rejections will happen during shutdown
+                @Override
+                public void rejectedExecution(Runnable runnable, ThreadPoolExecutor threadPoolExecutor) {
+                    // Log and discard
+                    logger.warn("Thread pool '{}' rejected execution of {}",
+                            new Object[] { poolName, runnable.getClass() });
+                    super.rejectedExecution(runnable, threadPoolExecutor);
+                }
+            });
+        }
+
+        public ExpressionThreadPoolExecutor(String threadPool, int corePoolSize, ThreadFactory threadFactory,
+                RejectedExecutionHandler rejectedHandler) {
+            super(corePoolSize, threadFactory, rejectedHandler);
+        }
+
+        @Override
+        protected void beforeExecute(Thread thread, Runnable runnable) {
+            super.beforeExecute(thread, runnable);
+            running.add(runnable);
+        }
+
+        @Override
+        protected void afterExecute(Runnable runnable, Throwable throwable) {
+            super.afterExecute(runnable, throwable);
+            running.remove(runnable);
+            futures.remove(runnable);
+            scheduled.remove(runnable);
+            if (throwable != null) {
+                Throwable cause = throwable.getCause();
+                if (cause instanceof InterruptedException) {
+                    // Ignore this, might happen when we shutdownNow() the executor. We can't
+                    // log at this point as the logging system might be stopped already.
+                    return;
+                }
+            }
+        }
+
+        public void schedule(final Runnable task, final Expression expression) {
+            if (task == null) {
+                throw new NullPointerException();
+            }
+
+            this.setCorePoolSize(this.getCorePoolSize() + 1);
```

Comment:
No general concern, but it does not really make much sense, since this thread will be blocked eternally and thus be effectively not be part of the pool anyhow, right? I am not sure if I understand your argument for using a thread in the pool. I think having a dedicated thread makes it easier to give it access to the queue (the ones from the pool should not know about the queue) and also to give it an explicit name, so that it is easily identified in the debugger.